### ISSUE TITLE
Spanner: Tests failing because of backup PITR

### ISSUE THREAD

**Issue by amanda-tarafa** [GoogleCloudPlatform/dotnet-docs-samples#1262]
Sample failures [here](https://source.cloud.google.com/results/invocations/575b363e-34b8-4655-aff3-cf969b13cdd7/targets/github%2Fdotnet-docs-samples%2Fspanner%2Fapi%2FSpanner.Samples.Tests%2FTestResults/tests). Seems like the version time is "in the future".

**Comment by skuruppu** [comment]
Hmmm that's interesting. We say create a backup at version_time = now. But I guess the clocks are not in sync so this can happen.

I don't have any great ideas on what time to set this to. We would ideally set the time in the past but we can't make any guarantees about when the database was created for the test.

**Comment by olavloite** [comment]
> Hmmm that's interesting. We say create a backup at version_time = now. But I guess the clocks are not in sync so this can happen.
> 
> I don't have any great ideas on what time to set this to. We would ideally set the time in the past but we can't make any guarantees about when the database was created for the test.

A `create_time` [property](https://cloud.google.com/spanner/docs/reference/rpc/google.spanner.admin.database.v1#database) was added to `Database` not so long ago, so I guess we can use that to set the `version_time`.

**Comment by amanda-tarafa** [comment]
It ominously says "if exists" :)

I think the problem is with Spanner not localizing dates (or this is what I remember from a fix from @olavloite in the libraries)? So maybe instead of using UTC, use the local date for the backup? I think the date is in the future because it's UTC, but it's treated as local in Spanner which is (for some) in the future? Not sure though if I'm remembering correctly.

**Comment by olavloite** [comment]
> It ominously says "if exists" :)
> 

Yup, that's 'very informative' ;-).
As far as I know, the 'if exists' boils down to: If the database was created after this field was made (publicly) available, it will be populated. So for test databases that are created now and in the future, we should be safe.

> I think the problem is with Spanner not localizing dates (or this is what I remember from a fix from @olavloite in the libraries)? So maybe instead of using UTC, use the local date for the backup? I think the date is in the future because it's UTC, but it's treated as local in Spanner which is (for some) in the future? Not sure though if I'm remembering correctly.

The other problem with UTC vs. local dates was in the hand-written client where parameters were bound to a statement. This seems to be in the auto-generated admin client. It could still be the same or a similar issue, so that's also worth looking into.

**Comment by jskeet** [comment]
I don't think this is a time zone issue - I think it's just clock skew. That's a problem with the sample - we shouldn't be encouraging users to specify "the current instant" when creating the backup, because it can so easily go wrong.

I think there are two use cases:

- "Create a backup with the current data" - don't specify the `VersionTime` property
- "Create a backup for a known-good timestamp" - specify that timestamp, e.g. from a commit

We could potentially have both of those as two separate samples, but we shouldn't have a sample which tries to be half way between the two.

**Comment by olavloite** [comment]
> I think there are two use cases:
> 
> * "Create a backup with the current data" - don't specify the `VersionTime` property
> * "Create a backup for a known-good timestamp" - specify that timestamp, e.g. from a commit
> 
> We could potentially have both of those as two separate samples, but we shouldn't have a sample which tries to be half way between the two.

I agree with the above. Having a simple sample for `CreateBackup` that does not take a `VersionTime` and a separate sample for creating a backup with a (valid) version time makes sense. It is however not in line with the current specs for the samples (and also not with the samples for the other client languages). @skuruppu What are your thoughts on this? Is this something that we should change for all the client libraries?

**Comment by skuruppu** [comment]
It is a little bit unfortunate that we had to modify the existing `CreateBackup` sample but it was done for a good reason. As you've probably noticed, creating a backup can take some time, often leading to tests timing out and being flaky. So instead of introducing another code sample, we decided to modify the existing sample.

I think the choice you made in #1263 is reasonable. It's also reasonable to make this consistent across other languages. But I will let @thiagotnunes chime in with his thoughts.
